### PR TITLE
Internal maneuver placement

### PR DESCRIPTION
**Pull request by dgearhart**
# Issue

Updated the maneuver placement at internal intersections when a turn starts with a straight internal edge.

## BEFORE
image

## AFTER
image

## Tasklist

 - [x] Add tests
 - [x] Add #fixes with the issue number that this PR addresses
 - [x] Generally use squash merge to rebase and clean comments before merging
 - [x] Update the [changelog](CHANGELOG.md)

### CODE REVIEW

File: test/gurka/test_instructions_internal_intersection.cc [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                          {{"mjolnir.admin",
+                            {VALHALLA_SOURCE_DIR "test/data/netherlands_admin.sqlite"}}});
+  }
+};
+
+gurka::map InstructionsInternalIntersection::map = {};
+
+///////////////////////////////////////////////////////////////////////////////
+// Left turn
+TEST_F(InstructionsInternalIntersection, BrokenLandParkway_TurnLeft_PatuxentWoodsDrive) {
```
Comment (danpaz) [inline test/gurka/test_instructions_internal_intersection.cc:61]: Can these tests also assert the begin/end location of the maneuvers?

File: src/odin/maneuversbuilder.cc [hunk 2]
```diff
@@ -2971,5 +2977,132 @@ void ManeuversBuilder::SetTraversableOutboundIntersectingEdgeFlags(std::list<Man
...
+
+  auto is_relative_straight = [](uint32_t turn_degree) -> bool {
+    return ((turn_degree >= kRelativeStraightTurnDegreeLowerBound) ||
+            (turn_degree <= kRelativeStraightTurnDegreeUpperBound));
+  };
+
+  Maneuver* prev_maneuver = nullptr;
+  for (auto& maneuver : maneuvers) {
+    // Verify that the previous maneuver is set
+    if (prev_maneuver) {
```
Comment (danpaz) [inline src/odin/maneuversbuilder.cc:3011]:
```suggestion
    if (!prev_maneuver) {
      prev_maneuver = &maneuver;
      continue;
    }
      
```

File: src/odin/maneuversbuilder.cc [hunk 3]
```diff
@@ -2971,5 +2977,132 @@ void ManeuversBuilder::SetTraversableOutboundIntersectingEdgeFlags(std::list<Man
...
+      // Skip destination maneuver
+      if (maneuver.IsDestinationType()) {
+        break;
+      }
+
+      // Maneuver must be a turn
+      if (is_turn_maneuver(maneuver.type())) {
```
Comment (danpaz) [inline src/odin/maneuversbuilder.cc:3018]:
```suggestion
      if (!is_turn_maneuver(maneuver.type())) {
        break;
      }
```

File: src/odin/maneuversbuilder.cc (hunk 2 above)
Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:3011]: this happen for each maneuver - would only want to update in one location

File: test/gurka/test_instructions_internal_intersection.cc (hunk 1 above)
Comment (dgearhart) [reply test/gurka/test_instructions_internal_intersection.cc:61]: you mean index instead of the distance?

Comment (danpaz) [reply test/gurka/test_instructions_internal_intersection.cc:61]: Right, yeah.

File: test/gurka/test_instructions_internal_intersection.cc [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  gurka::assert::raw::expect_instructions_at_maneuver_index(
+      result, ++maneuver_index, "Turn left onto Patuxent Woods Drive.",
+      "Turn left onto Patuxent Woods Drive.",
+      "Turn left onto Patuxent Woods Drive. Then You will arrive at your destination.",
+      "Continue for 40 meters.");
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Left u-turn
+TEST_F(InstructionsInternalIntersection, BrokenLandParkway_UturnLeft_BrokenLandParkway) {
```
Comment (mandeepsandhu) [inline test/gurka/test_instructions_internal_intersection.cc:90]: Are there other tests that would exercise a route that does not take an internal intersection? Just so we are sure we haven't changed anything for those scenarios? Rest LGTM.

File: test/gurka/test_instructions_internal_intersection.cc [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+using namespace valhalla;
+
+class InstructionsInternalIntersection : public ::testing::Test {
+protected:
+  static gurka::map map;
+
+  static void SetUpTestSuite() {
+    constexpr double gridsize_metres = 10;
+
+    const std::string ascii_map = R"(
```
Comment (mandeepsandhu) [inline test/gurka/test_instructions_internal_intersection.cc:17]: Do we want to add some degenerate cases like start and/or ending on an internal intersection, eg `A->F`, `B->G` etc?

File: src/odin/maneuversbuilder.cc (hunk 3 above)
Comment (dgearhart) [reply src/odin/maneuversbuilder.cc:3018]: would have to change the loop for this to work

File: test/gurka/test_instructions_internal_intersection.cc (hunk 4 above)
Comment (dgearhart) [reply test/gurka/test_instructions_internal_intersection.cc:90]: covered scenarios with RAD - will see about adding more into gurka

File: test/gurka/test_instructions_internal_intersection.cc (hunk 5 above)
Comment (dgearhart) [reply test/gurka/test_instructions_internal_intersection.cc:17]: sure - can add some more

File: test/gurka/test_instructions_internal_intersection.cc (hunk 1 above)
Comment (dgearhart) [reply test/gurka/test_instructions_internal_intersection.cc]: @danpaz I added index tests

File: test/gurka/test_instructions_internal_intersection.cc (hunk 5 above)
Comment (dgearhart) [reply test/gurka/test_instructions_internal_intersection.cc:19]: @mandeepsandhu I added test cases

File: test/gurka/gurka.h [hunk 6]
```diff
@@ -830,6 +830,30 @@ void expect_maneuvers(const valhalla::Api& result,
...
 
+/**
+ * Tests whether the expected sequence of maneuver begin path indexes is emitted for the route.
+ * Looks at the output of Odin in the result.
+ *
+ * @param result the result of a /route or /match request
+ * @param expected_indexes all the maneuver begin path indexes expected in the DirectionsLeg
+ *                         for the route
+ */
+void expect_maneuver_begin_path_indexes(const valhalla::Api& result,
```
Comment (danpaz) [inline test/gurka/gurka.h:841]: Nice addition 👍

**Review by danpaz — APPROVED** [review]

**Review by mandeepsandhu — DISMISSED** [review]
Thanks for adding the extra tests! 👍

**Review by mandeepsandhu — APPROVED** [review]

**Pull request merged**